smartsim/_core/_install/builder.py

```diff
@@ -385,10 +395,21 @@ def __init__(
         self.libtf_dir = libtf_dir
         self.torch_dir = torch_dir
 
-        # TODO: It might be worth making these constructor args so that users
-        #       of this class can configure exactly _what_ they are building.
-        self._os = OperatingSystem.from_str(platform.system())
-        self._architecture = Architecture.from_str(platform.machine())
+        # Sanity checks
+        self._check_backends_arm64()
+
+    def _check_backends_arm64(self) -> None:
+        if self._architecture == Architecture.ARM64:
+            unsupported = []
+            if self.build_tf:
+                unsupported.append("Tensorflow")
+            if self.build_onnx:
+                unsupported.append("ONNX")
+            if unsupported:
+                raise BuildError(
+                    f"The {'.'.join(unsupported)} backends are not "
+                    "supported on ARM64. Run with `smart build --no_tf`"
+                )
```

Comment:
I like the decision to try and move as much of the validation of dependencies into the constructor to error out early!

But I don't necessarily love the existence of this method. It's going to become very difficult to maintain if we were to expand out into different architectures w/o full dependency support and/or we want to become more prescriptive in our builds (e.g. different versions of libraries for different architectures, etc.)

What would you think of, instead, expanding the `_RAIBuildDependency` interface so that it included an `__is_satisfiable__` method:
```py
class _RAIBuildDependency(ABC):
    ...

    @abstractmethod
    def __is_satisfiable__(self) -> bool: ...
```

which can then be implemented by all dependencies that implement the interface:
```py
class _TFArchive(_WebTGZ, _RAIBuildDependency):
    ...

    def __is_satisfiable__(self) -> bool:
        try:
            self.url
        except BuildError:
            return False
        else:
             return True

# And presumably very similar implementations 
# for `_PTArchiveLinux`, `_PTArchiveMacOSX`, and `_ORTArchive`
```

and then checking for dependency satisfiability in `RedisAIBuilder.__init__` with a simple:
```py
class RedisAIBuilder(Builder):
    ...

    def __init__(self, ...) -> None:
        ...
        if not all(
            dep.__is_satisfiable__() for dep
            in sequence_of_deps_that_the_rai_builder_needs
        ):
            raise BuildError(f"{type(self).__name__} failed to satisfy "
                              "dependencies for this platform")
            # We can workshop the error message if want to be a bit 
            # more prescriptive, but you get the idea!
```

Reply:
Per conversation yesterday, this solution is not viable as we do not have all the information necessary at `__init__` time to construct the dependencies (e.g. we do not know if we are building for GPU), which should definitely be changed, but is also definitely not in the scope of this PR.

Switching to having all `_RAIBuildDependency`s to have a static method to return platform/architecture combinations that they can be built for. This will move the platform checks to each dependency which will be at least slightly more maintainable until we can move more build details into the `RedisAIBuilder.__init__`.